docs/osbuild.1.rst
=======
osbuild
=======

----------------------------------------------
Build-Pipelines for Operating System Artifacts
----------------------------------------------

:Manual section: 1
:Manual group: User Commands

SYNOPSIS
========

| ``osbuild`` [ OPTIONS ] MANIFEST
| ``osbuild`` [ OPTIONS ] -
| ``osbuild`` ``--help``
| ``osbuild`` ``--version``

DESCRIPTION
===========

**osbuild** is a build-system for operating system artifacts. It takes an input
manifest describing the build pipelines and produces file-system trees, images,
or other artifacts as output. Its pipeline description gives comprehensive
control over the individual steps to execute as part of a pipeline. **osbuild**
provides isolation from the host system as well as caching capabilities, and
thus ensures that pipeline builds will be deterministic and efficient.

OPTIONS
=======

**osbuild** reads the pipeline description from the file passed on the
command-line. To make **osbuild** read the pipeline description from standard
input, pass ``-``.

The following command-line options are supported. If an option is passed, which
is not listed here, **osbuild** will deny startup and exit with an error.

-h, --help                      print usage information and exit immediately
--version                       print version information and exit immediately
--cache=DIR, --store=DIR        directory where intermediary sources and file
                                system trees are stored (default: /var/cache/osbuild
                                when running as root, .osbuild otherwise)
-l DIR, --libdir=DIR            directory containing stages, assemblers, and
                                the osbuild library
--cache-max-size=SIZE           maximum size of the cache (bytes) or 'unlimited'
                                for no restriction (size may include an optional
                                unit suffix, like kB, kiB, MB, MiB and so on)
--checkpoint=CHECKPOINT         stage to commit to the object store during
                                build (can be passed multiple times)
--export=OBJECT                 object to export (can be passed multiple times)
--json                          output results in JSON format
--output-directory=DIR          directory where result objects are stored
--inspect                       return the manifest in JSON format including
                                all the ids
--monitor=TYPE                  name of the monitor to be used
--monitor-fd=NUM                file-descriptor to be used for the monitor
--stage-timeout                 set the maximal time (in seconds) each stage is
                                allowed to run
--break, --break=ID             open debug shell when executing stages; accepts
                                stage name or id (from --inspect) or * (for all)
--quiet, -q                     suppress normal output

NB: If neither ``--output-directory`` nor ``--checkpoint`` is specified, no
attempt to build the manifest will be made.

MANIFEST
========

The input to **osbuild** is a description of the pipelines to execute, as well
as required parameters to each pipeline stage. This data must be *JSON*
encoded. It is read from the file specified on the command-line, or, if ``-``
is passed, from standard input.

The format of the manifest is described in ``osbuild-manifest``\(5). The formal
schema of the manifest is available online as the OSBuild JSON Schema [#]_.

MONITOR
=======

Live activity of the pipeline execution can be monitored via a built-in set
of different monitors. The ``--monitor=TYPE`` option selects the type of
monitor that is used. Valid types are:

``NullMonitor``
        No live coverage is reported and all monitoring features are disabled.
        This is the default monitor if ``--json`` was specified on the
        command-line.
``LogMonitor``
        A human-readable live monitor of the pipeline execution. This monitor
        prints pipeline names, stage names, and relevant options of each stage
        as it is executed. Additionally, timing information is provided for
        each stage. The output is not machine-readable and is interspersed
        with the individual log messages of the stages.
        This is the default monitor if ``--json`` was **not** specified.
``JSONSeqMonitor``
        A json-sequence (RFC7464) based live monitor of the pipeline execution.
	Each line contains valid json and a record-separator (0x1E) at the
	end. The json consists of a `progress` object that contains a
	`context` object describes the context of the progress and details
	about the progress. Progress can be nested and a progress object
	may contain sub-progress. In addition each progress can emit
	`message` strings that contain detailed log information from
	the stage. An example:

	| {
	|   "message": "Starting module org.osbuild.grub2.inst",
	|   "context": {
	|     "origin": "osbuild.monitor",
	|     "pipeline": {
	|       "name": "image",
	|       "id": "b3e63a8ef3cd30be244ffe1749af05783014bf2c61b2e45f09bb6df0ef122ad8",
	|       "stage": {
	|         "name": "org.osbuild.grub2.inst",
	|         "id": "b3e63a8ef3cd30be244ffe1749af05783014bf2c61b2e45f09bb6df0ef122ad8"
	|       }
	|     },
	|     "id": "747c04fa9265e7de79ac5efae43c62375f4b6bb778cb24abd7963dc181139176"
	|   },
	|   "progress": {
	|     "name": "pipelines/sources",
	|     "total": 4,
	|     "done": 3,
	|     "progress": {
	|       "name": "stages",
	|       "total": 7,
	|       "done": 6
	|     }
	|   },
	|   "timestamp": 1710318022.3872378
	| }

Monitor output is written to the file-descriptor provided via
``--monitor-fd=NUM``. If none was specified, standard output is used.

OUTPUT
======

OSBuild only ever builds the requested artifacts, rather than all artifacts
defined in a manifest. Each stage and pipeline has an associated ID (which can
be acquired by passing ``--inspect``). To export an artifact after a stage or
pipeline finished, pass its ID via ``--export=ID``. A sub-directory will be
created in the output-directory with the ID as the name. The contents of the
artifact are then stored in that sub-directory.

Additionally, any completed pipeline or stage can be cached to avoid rebuilding
them in subsequent invocations. Use ``--checkpoint=ID`` to request caching of a
specific stage or pipeline.

EXAMPLES
========

The following sub-sections contain examples on running **osbuild**. Generally,
**osbuild** must be run with superuser privileges, since this is required to
create file-system images.

Example 1: Run an empty pipeline
--------------------------------

To verify your **osbuild** setup, you can run it on an empty pipeline which
produces no output:

    |
    | # echo {} | osbuild -
    |

Example 2: See pipeline IDs of a Fedora qcow2 image
---------------------------------------------------

To inspect a basic qcow2 image of Fedora, use:

    |
    | # osbuild ./samples/fedora-boot.json
    |

The pipeline definition ``./samples/fedora-boot.json`` is provided in the
upstream source repository of **osbuild**.

This will print out the pipeline IDs of the provided manifest but will
not actually build anything because no artifact was requested. The pipeline
IDs can then be passed to ``--export=`` to actually get the requested
artifacts.


Example 3: Build a Fedora qcow2 image
-------------------------------------

To build a basic qcow2 image of Fedora, use:

    |
    | # osbuild --output-dir ./out --export image ./samples/fedora-boot.json
    |

Note that the command requires to be run as root. It will create an out/image
directory and put the generated "disk.img" there.

Example 4: Run from a local checkout
------------------------------------

To run **osbuild** from a local checkout, use:

    |
    | # python3 -m osbuild --libdir . --output-dir ./out --export image samples/fedora-boot.json
    |

This will make sure to execute the **osbuild** module from the current
directory, as well as use it to search for stages, assemblers, and more.

SEE ALSO
========

``osbuild-manifest``\(5), ``osbuild-composer``\(1)

NOTES
=====

.. [#] OSBuild JSON Schema v2:
       https://osbuild.org/schemas/osbuild2.json


NEWS.md
## CHANGES WITH 39:

  * document osbuild and selinux integration
  * improvements to the CI

Contributions from: Achilleas Koutsou, Christian Kellner, Sanne Raymaekers,
                    Tomas Hozza

— Grenoble, 2021-10-06

## CHANGES WITH 38:

  * host services: make file descriptor passing more robust by checking the fds
    before sending them.

  * sources: fix a bug introduced in osbuild 37, where osbuild would fail to communicated with
    the sources host service for certain manifests with very large lists of items

Contributions from: Christian Kellner, Pierre-Yves Chibon

— Vienna, 2021-09-24

## CHANGES WITH 37:

  * Port sources to the host service infrastructure
  * `stages/org.osbuild.ostree.config`: add `bootloader` config option
  * `stages/org.osbuild.sysctld`: new stage for setting kernel parameters via sysctl.d
  * `stages/org.osbuild.pam.limits.conf`: new stage for configuring `pam_limits` module

Contributions from: Achilleas Koutsou, Alexander Larsson, Christian Kellner, Diaa Sami,
                    Martin Sehnoutka, Pierre-Yves Chibon, Sanne Raymaekers, Simon Steinbeiss,
                    Tomas Hozza

— Brno, 2021-09-22

## CHANGES WITH 36:

  * `stages/org.osbuild.selinux.config`: new stage for configuring SELinux on the system
  * `stages/org.osbuild.dnf.config`: new stage for changing persistent DNF configuration
  * `stages/org.osbuild.tuned`: new stage for setting active TuneD profile
  * `stages/org.osbuild.kickstart`: set passwords with --iscrypted
  * `stages/org.osbuild.tmpfilesd`: new stage for configuring tmpfiles.d
  * `util/rhsm`: Check if repositories is None before iterating

Contributions from: Achilleas Koutsou, Sanne Raymaekers, Tomas Hozza

— Berlin, 2021-09-08

## CHANGES WITH 35:
  * Bug fix release.

Contributions from: Christian Kellner

— Oslo, 2021-08-29

## CHANGES WITH 34:

  * `stages/bootiso.mono`: support for aarch64

  * `stages/ostree.{pull,deploy}`: support for specifying ostree
    remotes.

— Oslo, 2021-08-29

## CHANGES WITH 33:

  * `stages/org.osbuild.grub2`: add support for greenboot, i.e.
    automatically rolling back bad updates, i.e. updates that do not
    boot successfully.

  * `stages/org.osbuild.gzip`: new stage to compress files via gzip.

Contributions from: Christian Kellner, Diaa Sami

— Vöcklabruck, 2021-08-25

## CHANGES WITH 32:

  * `stages/org.osbuild.grub2`: allow the unified grub config scheme
    to be configured even in pure UEFI configurations.

Contributions from: Christian Kellner

- Asker, 2021-08-24

## CHANGES WITH 31:

  * **LVM2 support** was added. For this new stages and host services
    were added:

  * Support for parent devices was added to the `devices` host
    service: a device can now specify a device it depends on. An
    example is an LVM2 volume on a partition. When loading, the
    devices are pre-sorted so that all dependencies are in the
    correct order.

  * `stages/org.osbuild.lvm2.create`: new stage to create logical
     volumes, the volume group and the physical volume in one
     operation.

  * `devices/org.osbuild.lvm2.lv`: new device host service to activate
     a LVM2 logical volume. The logical volume is identified via the
     underlying device, so there is no need to know the logical
     volume's name or uuid.

  * `stages/org.osbuild.lvm2.metadata`: new stage to modify LVM2
     metadata.

  * `stages/org.osbuild.untar`: new stage that can be used to extract
    a tarball.

  * `runners/org.osbuild.rhel86`: new runner for RHEL 8.6, based on
    the current RHEL 8.2 runner.

  * `stages/org.osbuild.nm.conf`: new stage to create NetworkManager
    configuration files. Currently only a subset of settings are
    supported.

  * `inputs/org.osbuild.ostree.checkout`: new input to checkout an
    OSTree commit so that stages can use its contents.

  * `stages/org.osbuild.ostree.passwd` new stage to merge the groups
    and passwords from one or more existing commits into the
    corresponding files in the tree. This is needed to make sure that
    uids/gids stay stable between commits.

  * `stages/org.osbuild.grub2.inst`: fix prefix for dos layouts: When
    the partition layout is `dos` or `mbr`, the correct name for it in
    the prefix is `msdos`.

  * Loop controller improvements: various new helper methods to close
    and flush the loop devices.

  * `devices/org.osbuild.loopback`: explicitly flush the device buffer
    cache and clear the fd to ensure that all data is properly written
    to the backing file.

Contributions from: Tomas Hozza, Martin Sehnoutka, Christian Kellner,
                    Ondřej Budai, Thomas Lavocat

— Berlin, 2021-08-18

## CHANGES WITH 30:

  * Add support for building **OSTree based raw images**. This
    means creating a raw image and deploying a commit in it. For
    this various stages have been added:

  * `stage/org.osbuild.ostree.init-fs`: initializes a file
    system in OSTree layout.

  * `stages/org.osbuild.ostree.init-os`: initializes a new
    state root with a given name.

  * `stages/org.osbuild.ostree.config`: allows changing the
    configuration for an OSTree repo.

  * `stages/org.osbuild.ostree.deploy`: deploy an existing
    commit.

  * `stages/org.osbuild.ostree.fillvar`: use `systemd-tmpfiles`
    to provision `/var` for a stateroot and deployment.

  * `stages/org.osbuild.ostree.remotes`: allows to configure
    remotes for an OSTree repository.

  * `stages/org.osbuild.ostree.selinux`: re-label (parts) of
    an OSTree deployment.

  * `stages/org.osbuild.fstab`: support for creating an fstab file in
    an OSTree deployment. NB: this is a temporary solution and will
    very likely be removed in the next release.

  * **Manifest-Pre-Processor**: the various mpp tools got unified into a
    single tool which is now shipped as `osbuild-mpp` in the `osbuild-tools`
    package. Additionally, it supports variable expansion via `mpp-vars`

osbuild.py

import hashlib
import json
import os
import shutil
import subprocess
import sys
import tempfile


__all__ = [
    "Assembler",
    "AssemblerFailed",
    "BuildRoot",
    "load",
    "Pipeline",
    "Stage",
    "StageFailed",
]


RESET = "\033[0m"
BOLD = "\033[1m"


libdir = os.path.dirname(__file__)
if not os.path.exists(f"{libdir}/stages"):
    libdir = f"{sys.prefix}/lib/osbuild"


class StageFailed(Exception):
    def __init__(self, name, returncode, output):
        self.name = name
        self.returncode = returncode
        self.output = output


class AssemblerFailed(Exception):
    def __init__(self, name, returncode, output):
        self.name = name
        self.returncode = returncode
        self.output = output


class tmpfs:
    def __init__(self, path="/run/osbuild"):
        self.root = tempfile.mkdtemp(prefix="osbuild-tmpfs-", dir=path)
        self.mounted = False
        try:
            subprocess.run(["mount", "-t", "tmpfs", "-o", "mode=0755", "tmpfs", self.root], check=True)
            self.mounted = True
        except subprocess.CalledProcessError:
            self.unmount()

    def unmount(self):
        if not self.root:
            return
        if self.mounted:
            subprocess.run(["umount", "--lazy", self.root], check=True)
        os.rmdir(self.root)
        self.root = None

    def __del__(self):
        self.unmount()

    def __enter__(self):
        return self.root

    def __exit__(self, exc_type, exc_value, exc_tb):
        self.unmount()


class BuildRoot:
    def __init__(self, path="/run/osbuild"):
        self.root = tempfile.mkdtemp(prefix="osbuild-buildroot-", dir=path)
        self.mounted = False
        try:
            subprocess.run(["mount", "-o", "bind,ro", "/", self.root], check=True)
            self.mounted = True
        except subprocess.CalledProcessError:
            self.unmount()
            raise

        # systemd-nspawn silently removes some characters when choosing a
        # machine name from the directory name. The only one relevant for
        # us is '_', because all other characters used by
        # TemporaryDirectory() are allowed. Replace it with 'L's
        # (TemporaryDirectory() only uses lower-case characters)
        self.machine_name = os.path.basename(self.root).replace("_", "L")

    def unmount(self):
        if not self.root:
            return
        if self.mounted:
            subprocess.run(["umount", "--lazy", self.root], check=True)
        os.rmdir(self.root)
        self.root = None

    def run(self, argv, binds=[], readonly_binds=[], **kwargs):
        """Runs a command in the buildroot.

        Its arguments mean the same as those for subprocess.run().
        """
        command = "/run/osbuild/" + os.path.basename(argv[0])
        return subprocess.run([
            "systemd-nspawn",
            "--quiet",
            "--as-pid2",
            "--link-journal=no",
            "--volatile=yes",
            "--property=DeviceAllow=/dev/loop-control rw",
            "--property=DeviceAllow=block-loop rw",
            "--property=DeviceAllow=block-blkext rw",
            f"--machine={self.machine_name}",
            f"--directory={self.root}",
            f"--bind={libdir}/osbuild-run:/run/osbuild/osbuild-run",
            *[f"--bind={b}" for b in binds],
            *[f"--bind-ro={b}" for b in [argv[0] + ":" + command, *readonly_binds]],
            "/run/osbuild/osbuild-run",
            command
            ] + argv[1:], **kwargs
        )

    def __del__(self):
        self.unmount()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, exc_tb):
        self.unmount()


def print_header(title, options, machine_name):
    print()
    print(f"{RESET}{BOLD}{title}{RESET} " + json.dumps(options or {}, indent=2))
    print("Inspect with:")
    print(f"\t# nsenter -a --wd=/root -t `machinectl show {machine_name} -p Leader --value`")
    print()


def _get_system_resources_from_etc(resources):
    for r in resources:
        if not r.startswith("/etc"):
            raise ValueError(f"{r} is not a resource in /etc/")
        if ":" in r:
            raise ValueError(f"{r} tries to bind to a different location")
    return resources


class Stage:
    def __init__(self, name, base, options, resources):
        m = hashlib.sha256()
        m.update(json.dumps(name, sort_keys=True).encode())
        m.update(json.dumps(base, sort_keys=True).encode())
        m.update(json.dumps(options, sort_keys=True).encode())
        m.update(json.dumps(resources, sort_keys=True).encode())

        self.id = m.hexdigest()
        self.name = name
        self.options = options
        self.resources = resources

    def run(self, tree, interactive=False, check=True):
        with BuildRoot() as buildroot:
            if interactive:
                print_header(f"{self.name}: {self.id}", self.options, buildroot.machine_name)

            args = {
                "tree": "/run/osbuild/tree",
                "options": self.options,
            }

            r = buildroot.run(
                [f"{libdir}/stages/{self.name}"],
                binds=[
                    f"{tree}:/run/osbuild/tree",
                    "/dev:/dev"
                ],
                readonly_binds=_get_system_resources_from_etc(self.resources),
                encoding="utf-8",
                input=json.dumps(args),
                stdout=None if interactive else subprocess.PIPE,
                stderr=subprocess.STDOUT
            )
            if check and r.returncode != 0:
                raise StageFailed(self.name, r.returncode, r.stdout)

            return {
                "name": self.name,
                "returncode": r.returncode,
                "output": r.stdout
            }


class Assembler:
    def __init__(self, name, options, resources):
        self.name = name
        self.options = options
        self.resources = resources

    def run(self, tree, output_dir=None, interactive=False, check=True):
        with BuildRoot() as buildroot:
            if interactive:
                print_header(f"Assembling: {self.name}", self.options, buildroot.machine_name)

            args = {
                "tree": "/run/osbuild/tree",
                "options": self.options,
            }

            binds = ["/dev:/dev"]
            if output_dir:
                os.makedirs(output_dir, exist_ok=True)
                binds.append(f"{output_dir}:/run/osbuild/output")
                args["output_dir"] = "/run/osbuild/output"

            r = buildroot.run(
                [f"{libdir}/assemblers/{self.name}"],
                binds=binds,
                readonly_binds=[f"{tree}:/run/osbuild/tree"] + _get_system_resources_from_etc(self.resources),
                encoding="utf-8",
                input=json.dumps(args),
                stdout=None if interactive else subprocess.PIPE,
                stderr=subprocess.STDOUT)
            if check and r.returncode != 0:
                raise AssemblerFailed(self.name, r.returncode, r.stdout)

            return {
                "name": self.name,
                "returncode": r.returncode,
                "output": r.stdout
            }


class Pipeline:
    def __init__(self, base=None):
        self.base = base
        self.stages = []
        self.assembler = None

    def add_stage(self, name, options={}, resources=[]):
        base = self.stages[-1].id if self.stages else self.base
        stage = Stage(name, base, options, resources)
        self.stages.append(stage)

    def set_assembler(self, name, options={}, resources=[]):
        self.assembler = Assembler(name, options, resources)

    def run(self, output_dir, objects=None, interactive=False, check=True):
        os.makedirs("/run/osbuild", exist_ok=True)
        if objects:
            os.makedirs(objects, exist_ok=True)
        elif self.base:
            raise ValueError("'objects' argument must be given when pipeline has a 'base'")

        results = {
            "stages": []
        }
        with tmpfs() as tree:
            if self.base:
                subprocess.run(["cp", "-a", f"{objects}/{self.base}/.", tree], check=True)

            for stage in self.stages:
                r = stage.run(tree, interactive, check)
                results["stages"].append(r)
                if r["returncode"] != 0:
                    results["returncode"] = r["returncode"]
                    return results

            if self.assembler:
                r = self.assembler.run(tree, output_dir, interactive, check)
                results["assembler"] = r
                if r["returncode"] != 0:
                    results["returncode"] = r["returncode"]
                    return results

            last = self.stages[-1].id if self.stages else self.base
            if objects and last:
                output_tree = f"{objects}/{last}"
                shutil.rmtree(output_tree, ignore_errors=True)
                os.makedirs(output_tree, mode=0o755)
                subprocess.run(["cp", "-a", f"{tree}/.", output_tree], check=True)

        results["returncode"] = 0
        return results


def load(description):
    pipeline = Pipeline(description.get("base"))

    for s in description.get("stages", []):
        pipeline.add_stage(s["name"], s.get("options", {}), s.get("systemResourcesFromEtc", []))

    a = description.get("assembler")
    if a:
        pipeline.set_assembler(a["name"], a.get("options", {}), a.get("systemResourcesFromEtc", []))

    return pipeline
